executor/runtime/docker/docker.go

```diff
@@ -665,126 +670,46 @@ func prepareNetworkDriver(parentCtx context.Context, cfg Config, c runtimeTypes.
+	err = jsonpb.Unmarshal(stdoutPipe, &result)
 	if err != nil {
-		log.WithError(err).Error("Unable to read JSON from allocate command")
-		return nil, fmt.Errorf("Unable to read json from pipe: %+v", err) // nolint: gosec
-	}
-	c.SetVPCAllocation(&vpcAllocation)
-
-	if !killTimer.Stop() {
-		err = errors.New("Kill timer fired. Race condition")
-		log.WithError(err).Error("Accidentally killed the allocation command, leaving us in a 'unknown' state")
-		return nil, err
-	}
-
-	if !vpcAllocation.Success {
-		// Kill the thing
-		killCh <- struct{}{}
-		log.WithField("error", vpcAllocation.Error).Error("VPC Configuration error")
-		if (strings.Contains(vpcAllocation.Error, "invalid security groups requested for vpc id")) ||
-			(strings.Contains(vpcAllocation.Error, "InvalidGroup.NotFound") ||
-				(strings.Contains(vpcAllocation.Error, "InvalidSecurityGroupID.NotFound")) ||
-				(strings.Contains(vpcAllocation.Error, "Security groups not found"))) {
+		return nil, fmt.Errorf("Could not read / deserialize JSON from assignment command: %w", err)
+	}
+	switch t := result.Result.(type) {
+	case *vpcapi.VPCToolResult_Error:
+		log.WithField("error", t.Error.Error).Error("VPC Configuration error")
+		if (strings.Contains(t.Error.Error, "invalid security groups requested for vpc id")) ||
+			(strings.Contains(t.Error.Error, "InvalidGroup.NotFound") ||
+				(strings.Contains(t.Error.Error, "InvalidSecurityGroupID.NotFound")) ||
+				(strings.Contains(t.Error.Error, "Security groups not found"))) {
 			var invalidSg runtimeTypes.InvalidSecurityGroupError
-			invalidSg.Reason = errors.New(vpcAllocation.Error)
+			invalidSg.Reason = errors.New(t.Error.Error)
 			return nil, &invalidSg
 		}
-		return nil, fmt.Errorf("vpc network configuration error: %s", vpcAllocation.Error)
+		return nil, fmt.Errorf("vpc network configuration error: %s", t.Error.Error)
+	case *vpcapi.VPCToolResult_Assignment:
+		assignment = t.Assignment
+	default:
+		return nil, fmt.Errorf("Unknown type: %t", t)
 	}
 
-	if vpcAllocation.Generation == nil {
-		err = errors.New("Unable to determine allocation generation")
-		log.WithError(err).Warn("Could not process allocation")
-		killCh <- struct{}{}
-		return nil, err
+	err = allocationCommand.Wait()
+	if err != nil {
+		return nil, fmt.Errorf("Allocation command exited with unexpected error: %w", err)
 	}
 
-	switch g := (*vpcAllocation.Generation); g {
-	case vpcTypes.V1:
-		return func() error {
-			ctx, cancel := context.WithTimeout(context.Background(), 45*time.Second)
-			defer cancel()
-
-			killCh <- struct{}{}
-			var err2 error
-			select {
-			case err2 = <-errCh:
-			case <-ctx.Done():
-				return errors.Wrap(ctx.Err(), "Error waiting for v1 assignment command to exit")
-			}
-			if err2 != nil {
-				log.WithError(err2).Error("Received error on termination of assignment command")
-				return errors.Wrap(err2, "Could not unassign task IP address")
-			}
-			return nil
-		}, nil
-	case vpcTypes.V3:
```

Comment:
Is your intent to remove *all* case statements when it comes to instance generation?
Or do you want to leave them in because we know we'll get a v4 some day?
Because in previous parts of the code you just took out the v1/v2 and wasn't sure if you intended to go "all the way" and remove all the case stuff.

Reply:
So, the idea its to move away from the switch statements in throughout the code, and add a pointer method on the Assignment struct that gives the specific info the user may be interested in.